### ISSUE TITLE
Apply core threshold correctly using 'quantile'

### ISSUE THREAD

**Issue by bockthom** [se-sic/coronet#205]
### Description

At the moment, determining core developers via a core threshold is computed somehow wrongly, at least, when using centrality values.

What we do is the following: We sum up all the "centrality values", determine the threshold by multiplying the core threshold with the computed sum to get a new threshold (see code snippet below), and afterwards sum up all centrality values starting from the largest centrality value up until we reach the threshold.
https://github.com/se-sic/coronet/blob/a7f4123dcb86a7a0f76adf791b66a7d9c24f7647/util-core-peripheral.R#L1148-L1159
https://github.com/se-sic/coronet/blob/a7f4123dcb86a7a0f76adf791b66a7d9c24f7647/util-core-peripheral.R#L1098-L1105

This works on count-based centrality values, but may lead to a wrong classification on centrality-based centrality values. Let's consider the following example:

**Count-based:** Let's have developers A, B, and C that made 2, 4, 4 commits. When using a core threshold of 0.8, then we get a threshold of (2+4+4)*0.8 = 8. Afterwards we sum up the commit counts beginning from the largest one until we reach 8. 4+4 = 8, so B and C are considered core. This example is correct: The 0.8 quantile of vector 2, 4, 4 is 4, telling us that all values >= 4 should be treated as core, which holds for B and C.

**Network-based:** Let's have developers A, B, C, and D. Let's assume their network-centrality values are 0.2, 0.4, 0.7, 0.8 (I am not sure if it is possible to end up in exactly these values in a real network, but let's assume that would work.) Then the summing-up procedure will end up in a threshold of (0.2+0.4+0.7+0.8)*0.8 = 1.68. When we afterwards sum up the centrality values to reach the threshold of 1.68, we see that we need 0.8+0.7+0.4 > 1.68. Hence B, C, and D would be treated as core developers.
However, when we just take the 0.8 quantile of vector (0.2, 0.4, 0.7, 0.8), we get a value of 0.74, resulting in only developer D being treated as core. So, in this case, our computation says that B, C, and D are core, whereas the quantile-computation only tells us that D is core...

### Suggested solution

Let's use `quantile(centrality.list, CORE.THRESHOLD)` to determine the threshold, and everything which is above this threshold should be considered core.

### Versions

All versions of coronet are affected, as this implementation part is part of coronet almost since the very beginning.

**Comment by clhunsen** [comment]
For me, the current implementation works well for count-based values (and cannot be swapped for the `quantile` implementation!) as we answer the question, "Which developers contribute (more than) 80 % of the commits/mails/..., in relation to the total number of commits/mails/... (beginning with the developers contributing most)?".
For the count-based example you give, the suggestion with `quantile` may (accidentally) work, but what about the vector `c(a = 2, b = 4, c = 7, d = 8)` (based on your network-based example above)? In this new example, a total of 21 commits are performed, 80 % of 21 commits are 16.8 commits. Beginning with the developers contributing most, Developers d, c, and b are the ones that contribute (more than) 80 % of the commits (8 + 7 + 4 = 19 >= 16.8). 
With `quantile(c(a = 2, b = 4, c = 7, d = 8), 0.8)`, you obtain a value of  `7.4` – resulting in only Developer d considered core as `8` is the only value greater than `7.4`. Obviously, a plausibility test falsifies this result: The 8 commits by Developer d represent, roughly, only 38 % of the total number of commits. Differently said, when choosing only Developer d with a value of `8 > 7.4` , we obtain a list of all developers that have performed more commits than 80 % of all developers have performed independently. But, this does not answer the questions stated above, does it?  Remember, the 0.5 quantile is the median: 50 % of the values are larger and smaller than this value, respectively. Meaning, the 0.8 quantile represents a value where 80 % of the considered values are smaller and 20 % are larger. This value cannot yield results for the question above, I assume.

For network-based centrality values, you need to answer whether there is something as "the sum of all centrality values" makes sense somehow. The total number of all commits make sense, but what is the sum of all centrality values?! This means, it may be plausible to say that only developers are considered core that have a centrality value above the 0.8 quantile of all centrality values (see your example above).

In the end, you may end up with two different threshold calculations, I guess. :thinking:

**Comment by bockthom** [comment]
Thanks for your reply @clhunsen. I totally agree with you. As our examples show, we need two different implementations for count-based and network-based classifications. Interestingly, this actually what is done in Codeface, what you can see if you follow the link we have added in `util-core-peripheral.R`:
https://github.com/se-sic/coronet/blob/a7f4123dcb86a7a0f76adf791b66a7d9c24f7647/util-core-peripheral.R#L25-L26

However, do you know why our implementation deviates from the one ours is derived from? In Codeface, there is a differentiation between when to use the sum-based approach and when to use the quantile-approach, but why didn't we only take the sum-based approach and used this approach also for network-based centrality computations? Our implementation was created about 4 years ago, but I don't know why we only had taken over parts of the Codeface implementation. I wonder why none of us had complained about using this implementation for network-based classifications...

**Comment by clhunsen** [comment]
> However, do you know why our implementation deviates from the one ours is derived from?

I do not know. Unfortunately, this decision is not documented in the repository and I cannot remember any disucssion on that... Maybe, we agreed on removing it as we do not need it actually – for whatever reason. Maybe, we started with the count-based classifications and never thought about adjusting the threshold calculation when we added the network-based classifications later. Unfortunately, I cannot find any evidence on both these ideas in the repo.  

> I wonder why none of us had complained about using this implementation for network-based classifications...

I cannot tell you, I have never used that kind of classification and trusted the existing implementation. :man_shrugging:
